src/hangar/records/hashs.py

```diff
@@ -13,68 +20,178 @@ def __init__(self, hashenv: lmdb.Environment):
 
     # ------------------ traversing the unpacked records ----------------------
 
-    def _traverse_all_hash_records(self):
-        """Pull out all records in the database as a tuple of binary encoded
+    def _traverse_all_hash_records(self, keys: bool = True, values: bool = True):
+        """PUll out all binary encoded data hash records.
 
-        Returns
+        Parameters
+        ----------
+        keys : bool, optional
+            if True, returns keys, by default True
+        values : bool, optional
+            if True, return values, by default True
+
+        Yields
         -------
-        list of tuples of bytes
-            list type stack of tuples with each db_key, db_val pair
+        Union[bytes, Tuple[bytes, bytes]]
+            bytes of keys if ``keys=True`` and ``values=False``
+
+            bytes of values if ``keys=False`` and ``values=True``
+
+            Tuple of bytes corresponding to ``(keys, values)`` if
+            ``keys=True`` and ``values=True``.
         """
         startHashRangeKey = f'{c.K_HASH}'.encode()
         try:
             hashtxn = TxnRegister().begin_reader_txn(self._hashenv)
             with hashtxn.cursor() as cursor:
                 hashsExist = cursor.set_range(startHashRangeKey)
-                while hashsExist:
-                    hashRecKey, hashRecVal = cursor.item()
-                    if hashRecKey.startswith(startHashRangeKey):
-                        yield hashRecKey
-                        hashsExist = cursor.next()
-                        continue
-                    else:
-                        hashsExist = False
+                # divide loop into returned type sections as perf optimization
+                # (rather then if/else checking on every iteration of loop)
```

Comment:
out of curiosity: Was this performance gain significant?

Reply:
Well technically there is no direct performance gain over the "old" method, since it didn't do the checks, but check the following benchmark for the actual saves we do see:

```python
iterlist = [(f'k_{i}', f'v_{i}') for i in range(100_000)]

def run_slow(keys: bool = True, values: bool = True):
    for db_k, db_v in iterlist:
        if (keys is True) and (values is True):
            yield (db_k, db_v)
        elif (keys is True) and (values is False):
            yield db_k
        elif (keys is False) and (values is True):
            yield db_v
        else:
            raise ValueError('both keys and values param cannot be false')
            
def run_fast(keys: bool = True, values: bool = True):
    if (keys is True) and (values is True):
        for db_k, db_v in iterlist:
            yield (db_k, db_v)
    elif (keys is True) and (values is False):
        for db_k, _ in iterlist:
            yield db_k
    elif (keys is False) and (values is True):
        for _, db_v in iterlist:
            yield db_v
    else:
        raise ValueError('both keys and values param cannot be false')
```
```python
%timeit list(run_slow(keys=True, values=True))
%timeit list(run_slow(keys=True, values=False))
%timeit list(run_slow(keys=False, values=True))
```

```
12.6 ms ± 374 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10.9 ms ± 289 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10.9 ms ± 204 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

```python
%timeit list(run_fast(keys=True, values=True))
%timeit list(run_fast(keys=True, values=False))
%timeit list(run_fast(keys=False, values=True))
```

```
10.4 ms ± 110 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.72 ms ± 193 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.76 ms ± 311 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```